### PR TITLE
Mjolnir would apply stale ACL to rooms during batching 

### PR DESCRIPTION
**Pull request by Gnuxie**
<del>I don't understand why this happens, just some Synapse behaviour.
In the test we create a bunch of server rules in a policy list in order to test "batching" of actions from rules.
The problem was that even though we wait for a server response to each event send (for each rule).
These rules do not necessarily appear in the room state of the policy list immediately after, which meant
we weren't banning as many servers as expected in protected rooms (some rules were missing from the list state).
This commit changes that by waiting until all the rules become visible in room state before
going forwards.</del>

Mjolnir will call `applyServerAcls` several times while a policy list is being updated, sometimes concurrently. This means a request to set a server ACL in a room which has old data can finish after a more recent recent request with the correct ACL. This means that the old ACL gets applied to the rooms (for a moment).  

This is a follow up from https://github.com/matrix-org/mjolnir/pull/314/commits/551065815e65e6117d7e8105afd6f5da9b87c8af

### CODE REVIEW

File: src/actions/ApplyAcl.ts [hunk 1]
```diff
@@ -31,6 +35,23 @@ import { ERROR_KIND_FATAL, ERROR_KIND_PERMISSION } from "../ErrorCache";
...
+    // So we naively promise chain them.
+    let timeoutId: NodeJS.Timeout|null = null;
+    inProgress = Promise.race([
+        new Promise((resolve, reject) => {
+            inProgress.finally(() => {
+                _applyServerAcls(lists, roomIds, mjolnir).then(resolve).catch(reject)
+            })
+        }),
+        // We don't want to stop ACL being applied if for some reason one gets stuck for 10 minutes.
+        new Promise((_, reject) => timeoutId = setTimeout(() => reject(new Error("Unable to apply ACL within a reasonable time.")), 600_000))
```
Comment (Yoric) [inline src/actions/ApplyAcl.ts:48]: Please make this a constant. Ideally, configurable but that can wait for a followup.

**Comment by Gnuxie** [comment]
using the same queue from mjolnir is kind of bad tbh, since applying ACL is critical

**Review by Yoric — APPROVED** [review]

**Comment by Gnuxie** [comment]
This is blocked by #334 probably, but generally there needs to be a queue in Mjolnir just for applyingACL that isn't throttled, it just needs to provide mutual exclusion for ApplyACL.

File: test/integration/banListTest.ts [hunk 2]
```diff
@@ -260,20 +260,23 @@ describe('Test: ACL updates will batch when rules are added in succession.', fun
         mjolnir.joinRoom(banListId);
         this.mjolnir!.watchList(Permalinks.forRoom(banListId));
         const acl = new ServerAcl(serverName).denyIpAddresses().allowServer("*");
-        for (let i = 0; i < 200; i++) {
+        const evilServerCount = 200;
```
[the remark this answers is not in the document]
Comment (Gnuxie) [reply test/integration/banListTest.ts:264]: not sure about this, it isn't in global scope and i'm not sure whether you'd do that even if it were

File: src/actions/ApplyAcl.ts [hunk 3]
```diff
@@ -31,6 +31,16 @@ import { ERROR_KIND_FATAL, ERROR_KIND_PERMISSION } from "../ErrorCache";
  * @param {Mjolnir} mjolnir The Mjolnir client to apply the ACLs with.
  */
 export async function applyServerAcls(lists: BanList[], roomIds: string[], mjolnir: Mjolnir): Promise<RoomUpdateError[]> {
+    // There's probably a name for what this actually is, but we only want one instance of `_applyServerAcls` to be running at any time.
+    return new Promise(resolve => {
+        mjolnir.taskQueue.push(async () => {
+            // We have to await rather than chain with .then because otherwise we will not block the task queue while we sync ACL.
+            resolve(await _applyServerAcls(lists, roomIds, mjolnir));
```
[the remark this answers is not in the document]
Comment (Gnuxie) [reply src/actions/ApplyAcl.ts]: changed to propagate the promise returned by `applyServerAcls`.

**Pull request merged**